_source/_posts/2020-12-16-react-login.md

```diff
@@ -0,0 +1,599 @@
+      email: '',
+      password: ''
+    };
+
+    this.oktaAuth = new OktaAuth({
+      issuer: props.baseUrl,
+      clientId: '{ClientId}',
+      pkce: true
+    });
+
+    this.handleEmailChange = this.handleEmailChange.bind(this);
+    this.handlePasswordChange = this.handlePasswordChange.bind(this);
+    this.signIn = this.signIn.bind(this);
+  }
+
+  handleEmailChange(event) {
+    this.setState({email: event.target.value});
+  }
+
+  handlePasswordChange(event) {
+    this.setState({password: event.target.value});
+  }
+
+  async signIn(event) {
+    event.preventDefault();
+    const transaction = await this.oktaAuth.signIn({
+      username: this.state.email,
+      password: this.state.password
+    });
+
+    if (transaction.status === 'SUCCESS') {
+      this.props.oktaAuth.signInWithRedirect({sessionToken: transaction.sessionToken})
+    } else {
+      throw new Error('Could not sign in: ' + transaction.status);
+    }
+  }
+
+  render() {
+    return (
+      <form onSubmit={this.signIn} className="login-form">
+        <h2>Log In</h2>
+        <p>Please login to continue</p>
+        <label className="full-width-input">
+          Email
+          <input type="text" placeholder="Email" value={this.state.email} onChange={this.handleEmailChange} required />
+        </label>
+        <label className="full-width-input">
+          Password
+          <input type="password" placeholder="Password" value={this.state.password} onChange={this.handlePasswordChange} required autoComplete="off" />
+        </label>
+        <button className="button">Login</button>
+      </form>
+    );
+  }
+};
+
+export default withOktaAuth(OktaSignInWidget);
+```
+
+You will notice a number of differences compared to the implementation in the previous section. First, the `render()` function now returns its own form with **Email** and **Password** input elements. The functions `handleEmailChange()` and `handlePasswordChange()` update the components state, always reflecting the value in the input fields.
+
+In the constructor, you will see that an `OktaAuth` object is created and assigned to the `oktaAuth` property. This object is used inside the `signIn()` function, which is called when the form is submitted. `OktaAuth` is responsible for connecting to the Okta servers, sending the user credentials, and in return, receiving a session token.
+
+The last thing you will notice is the fact that the `OktaSignInWidget` class is wrapped in the `withOktaAuth()` function before being exported. This adds the `authService` to the properties of the object. Remember, this widget is rendered within the `Security` widget. Inside the `signIn()` function the `authService` is used to set the application's session token and redirect the user to the protected page they wanted to visit before the sign-in was triggered.
+
+To round things up, add a bit of styling for the login form. Open `src/App.css` and paste the following code at the end of the file.
```

Comment:
To round things "up" or "out"? https://www.merriam-webster.com/dictionary/round%20out

Reply:
Will fix.